File: 06_breadth-first_search/python/my_bfs.py
#!/usr/bin/python3

import functools

class Cities:
    def __init__(self, city_roads):
        self.city_roads = city_roads
        self.city_list = list(city_roads.keys())

    def __del__(self):
        self.city_roads = None
        self.city_list = []

    def cities(self):
        return self.city_list

    def neighbors(self, city):
        if city in self.city_list:
            return list(self.city_roads[city].keys())
        else:
            return []

    def neighbor_distance(self, city, city2):
        if city == city2:
            return 0
        elif city in self.city_list and city2 in self.city_roads[city].keys():
            return self.city_roads[city][city2]
        else:
            return False

    def distance(self, city_list):
        l0 = len(city_list)
        if l0 <= 1:
            return 0
        else:
            (dist, _) = functools.reduce(
                lambda x, y: (x[0] + self.neighbor_distance(x[1], y), y),
                city_list,
                (0, city_list[0])
            )
            return dist

    def _get_roads(self, travelled_path, target_city):
        last_city = travelled_path[-1]
        if last_city == target_city:
            return [travelled_path]
        else:
            neighbors = self.neighbors(last_city)
            next_cities = [city for city in neighbors if not city in travelled_path]
            return \
                functools.reduce(
                    lambda acc_roads, next_city: acc_roads + self._get_roads(travelled_path + [next_city], target_city),
                    next_cities,
                    []
                )

    def roads(self, city, city2):
        return self._get_roads([city], city2)

    def distance_of_roads(self, roads):
        for road in roads:
            print(road, self.distance(road))
        return None

CityRoads = {
    "Y":{"A": 4, "B": 5, "C": 2},
    "A":{"P": 3, "Y": 4},
    "B":{"P": 1,"Y": 5, "N": 3},
    "C":{"J": 6, "T": 10, "Y": 2},
    "J":{"C": 6},
    "T":{"C": 10},
    "P":{"A": 3, "B": 1},
    "N":{"B": 5}
}

CRS = Cities(CityRoads)
print(CRS.cities())
print(CRS.neighbors("Y"))
print(CRS.neighbor_distance("Y","B"))
print(CRS.distance(["Y", "B", "P", "A"]))
print(CRS.roads("Y", "A"))
print(CRS.distance_of_roads(CRS.roads("Y", "A")))

File: 06_breadth-first_search/erlang/my_bfs.erl
-module(my_bfs).

-author("Song Feng").

%% APIs
-export([main/0]).


main() ->
    CityRoads = #{
        "Y" => [{"A", 4}, {"B", 5}, {"C", 2}],
        "A" => [{"P", 3}, {"Y", 4}],
        "B" => [{"P", 1}, {"Y", 5}, {"N", 3}],
        "C" => [{"J", 6}, {"T", 10}, {"Y", 2}],
        "J" => [{"C", 6}],
        "T" => [{"C", 10}],
        "P" => [{"A", 3}, {"B", 1}],
        "N" => [{"B", 5}]
    },
    io:format("~p~n", [cities(CityRoads)]),
    io:format("~p~n", [neighbors(CityRoads, "Y")]),
    io:format("~p~n", [neighbor_distance(CityRoads, "Y", "B")]),
    io:format("~p~n", [distance(CityRoads, ["Y", "B", "P"])]),
    io:format("------------------------------~n"),
    io:format("~p~n", [get_roads(CityRoads, ["Y"], "B")]).


cities(CRS) ->
    maps:keys(CRS).

neighbors(CRS, City) ->
    IsInvalidCity = maps:is_key(City, CRS),
    if IsInvalidCity ->
            {ok, NeighborRoads} = maps:find(City, CRS),
            [Neighbor || {Neighbor, _} <- NeighborRoads];
        true ->
            []
    end.

neighbor_distance(_CRS, City, City) -> 0;

neighbor_distance(CRS, City, City2) ->
    case maps:find(City, CRS) of
        {ok, NeighborRoads} ->
            {City2, Dist} = lists:keyfind(City2, 1, NeighborRoads),
            Dist;
        _ ->
            false
    end.

distance(CRS, CityPath) ->
    if length(CityPath) =< 1 ->
            0;
        true ->
            {_, Dist} = lists:foldl(
                fun (X, {Last, Acc}) -> {X, Acc + neighbor_distance(CRS, Last, X)} end,
                {hd(CityPath), 0},
                CityPath
            ),
            Dist
    end.

get_roads(CRS, TravelledPath, TargetCity) ->
    LastCity = hd(lists:reverse(TravelledPath)),
    case LastCity of
        TargetCity ->
            [TravelledPath];
        _ ->
            Neighbors = neighbors(CRS, LastCity),
            NextCities = [City || City <- Neighbors, not lists:member(City, TravelledPath)],
            case NextCities of
                [] ->
                    [];
                _ ->
                    lists:foldl(
                        fun (NextCity, AccRoads) -> AccRoads ++ get_roads(CRS, TravelledPath ++ [NextCity], TargetCity) end,
                        [],
                        NextCities)
            end
    end.
